File: teacher/lib/data/services/live_session_service.dart
import 'dart:async';
import 'dart:math';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

/// Firestore service for Live Lesson sessions (teacher side).
/// Mirrors the web `live-session.service.ts` using direct Firestore access.
class LiveSessionService {
  final _db = FirebaseFirestore.instance;
  final _auth = FirebaseAuth.instance;

  // ── Helpers ──

  String _generateJoinCode() {
    const chars = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
    final rng = Random.secure();
    return List.generate(6, (_) => chars[rng.nextInt(chars.length)]).join();
  }

  String get _uid => _auth.currentUser!.uid;
  String get _displayName =>
      _auth.currentUser?.displayName ?? 'Teacher';

  // ── Session CRUD ──

  Future<String> createSession({
    required String lessonId,
    required String lessonTitle,
    required String organizationId,
  }) async {
    final now = DateTime.now().toIso8601String();
    final joinCode = _generateJoinCode();

    final docRef = await _db.collection('liveSessions').add({
      'lessonId': lessonId,
      'lessonTitle': lessonTitle,
      'organizationId': organizationId,
      'teacherId': _uid,
      'teacherName': _displayName,
      'status': 'active',
      'joinCode': joinCode,
      'currentSlideIndex': 0,
      'focusMode': false,
      'participantCount': 1,
      'createdAt': now,
      'updatedAt': now,
    });

    // Add teacher as first participant
    await docRef.collection('participants').doc(_uid).set({
      'userId': _uid,
      'name': _displayName,
      'role': 'teacher',
      'isOnline': true,
      'joinedAt': now,
      'lastActiveAt': now,
    });

    return docRef.id;
  }

  Future<void> endSession(String sessionId) async {
    await _db.collection('liveSessions').doc(sessionId).update({
      'status': 'ended',
      'updatedAt': DateTime.now().toIso8601String(),
    });
  }

  Future<Map<String, dynamic>?> getSession(String sessionId) async {
    final doc = await _db.collection('liveSessions').doc(sessionId).get();
    if (!doc.exists) return null;
    return {'id': doc.id, ...doc.data()!};
  }

  Future<Map<String, dynamic>?> findActiveSessionForLesson(
      String lessonId) async {
    final snap = await _db
        .collection('liveSessions')
        .where('lessonId', isEqualTo: lessonId)
        .where('status', isEqualTo: 'active')
        .limit(1)
        .get();
    if (snap.docs.isEmpty) return null;
    final d = snap.docs.first;
    return {'id': d.id, ...d.data()};
  }

  // ── Real-time Subscriptions ──

  Stream<Map<String, dynamic>?> watchSession(String sessionId) {
    return _db
        .collection('liveSessions')
        .doc(sessionId)
        .snapshots()
        .map((snap) {
      if (!snap.exists) return null;
      return {'id': snap.id, ...snap.data()!};
    });
  }

  Stream<List<Map<String, dynamic>>> watchParticipants(String sessionId) {
    return _db
        .collection('liveSessions/$sessionId/participants')
        .snapshots()
        .map((snap) => snap.docs.map((d) => d.data()).toList());
  }

  Stream<List<Map<String, dynamic>>> watchReactions(String sessionId) {
    return _db
        .collection('liveSessions/$sessionId/reactions')
        .orderBy('createdAt', descending: true)
        .limit(20)
        .snapshots()
        .map((snap) =>
            snap.docs.map((d) => {'id': d.id, ...d.data()}).toList());
  }

  // ── Annotations ──

  Future<void> addAnnotation(
    String sessionId, {
    required String type,
    required List<Map<String, double>> points,
    required String color,
    required double width,
    int slideIndex = 0,
  }) async {
    await _db.collection('liveSessions/$sessionId/annotations').add({
      'sessionId': sessionId,
      'type': type,
      'points': points,
      'color': color,
      'width': width,
      'slideIndex': slideIndex,
      'authorId': _uid,
      'createdAt': DateTime.now().toIso8601String(),
    });
  }

  Future<void> clearAnnotations(String sessionId) async {
    final snap =
        await _db.collection('liveSessions/$sessionId/annotations').get();
    final batch = _db.batch();
    for (final doc in snap.docs) {
      batch.delete(doc.reference);
    }
    await batch.commit();
  }

  // ── Cursor (Throttled) ──

  Timer? _cursorTimer;
  Map<String, double>? _pendingCursor;

  void updateCursor(String sessionId, double x, double y) {
    _pendingCursor = {'x': x, 'y': y};
    if (_cursorTimer != null) return;

    _cursorTimer = Timer(const Duration(milliseconds: 100), () async {
      if (_pendingCursor != null) {
        try {
          await _db
              .collection('liveSessions/$sessionId/participants')
              .doc(_uid)
              .update({
            'cursorX': _pendingCursor!['x'],
            'cursorY': _pendingCursor!['y'],
            'lastActiveAt': DateTime.now().toIso8601String(),
          });
        } catch (_) {}
      }
      _cursorTimer = null;
      _pendingCursor = null;
    });
  }

  void disposeCursorThrottle() {
    _cursorTimer?.cancel();
    _cursorTimer = null;
    _pendingCursor = null;
  }

  // ── Reactions ──

  Future<void> addReaction(String sessionId, String type) async {
    await _db.collection('liveSessions/$sessionId/reactions').add({
      'sessionId': sessionId,
      'userId': _uid,
      'userName': _displayName,
      'type': type,
      'createdAt': DateTime.now().toIso8601String(),
    });
  }

  // ── Kick ──

  Future<void> kickParticipant(String sessionId, String userId) async {
    await _db
        .collection('liveSessions/$sessionId/participants')
        .doc(userId)
        .delete();
  }
}

File: src/services/live-session.service.ts
import {
  collection, doc, getDoc, getDocs, addDoc, updateDoc,
  deleteDoc, query, where, orderBy, onSnapshot, setDoc,
  limit, getCountFromServer, writeBatch
} from 'firebase/firestore';
import { db } from '../lib/firebase';
import type { LiveSession, LiveParticipant, LiveAnnotation, LiveReaction } from '../types';

// ---- Helpers ----

const generateJoinCode = (): string => {
  const chars = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789'; // no I/O/0/1 for readability
  let code = '';
  for (let i = 0; i < 6; i++) {
    code += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return code;
};

// ---- Session CRUD ----

export const createLiveSession = async (
  lessonId: string,
  lessonTitle: string,
  organizationId: string,
  teacherId: string,
  teacherName: string
): Promise<string> => {
  const joinCode = generateJoinCode();
  const now = new Date().toISOString();

  const docRef = await addDoc(collection(db, 'liveSessions'), {
    lessonId,
    lessonTitle,
    organizationId,
    teacherId,
    teacherName,
    status: 'active',
    joinCode,
    currentSlideIndex: 0,
    focusMode: false,
    participantCount: 0,
    createdAt: now,
    updatedAt: now,
  });

  // Add teacher as first participant
  await setDoc(doc(db, `liveSessions/${docRef.id}/participants`, teacherId), {
    userId: teacherId,
    name: teacherName,
    role: 'teacher',
    isOnline: true,
    joinedAt: now,
    lastActiveAt: now,
  });

  return docRef.id;
};

export const endLiveSession = async (sessionId: string): Promise<void> => {
  await updateDoc(doc(db, 'liveSessions', sessionId), {
    status: 'ended',
    updatedAt: new Date().toISOString(),
  });
};

export const getLiveSession = async (sessionId: string): Promise<LiveSession | null> => {
  const d = await getDoc(doc(db, 'liveSessions', sessionId));
  if (!d.exists()) return null;
  return { id: d.id, ...d.data() } as LiveSession;
};

export const findSessionByCode = async (code: string): Promise<LiveSession | null> => {
  const q = query(
    collection(db, 'liveSessions'),
    where('joinCode', '==', code.toUpperCase()),
    where('status', '==', 'active'),
    limit(1)
  );
  const snap = await getDocs(q);
  if (snap.empty) return null;
  const d = snap.docs[0];
  return { id: d.id, ...d.data() } as LiveSession;
};

export const findActiveSessionForLesson = async (lessonId: string): Promise<LiveSession | null> => {
  const q = query(
    collection(db, 'liveSessions'),
    where('lessonId', '==', lessonId),
    where('status', '==', 'active'),
    limit(1)
  );
  const snap = await getDocs(q);
  if (snap.empty) return null;
  const d = snap.docs[0];
  return { id: d.id, ...d.data() } as LiveSession;
};

// ---- Participants ----

const updateParticipantCount = async (sessionId: string) => {
  const coll = collection(db, `liveSessions/${sessionId}/participants`);
  const snap = await getCountFromServer(coll);
  await updateDoc(doc(db, 'liveSessions', sessionId), {
    participantCount: snap.data().count,
  });
};

export const joinLiveSession = async (
  sessionId: string,
  userId: string,
  name: string,
  avatarUrl?: string
): Promise<void> => {
  const now = new Date().toISOString();
  await setDoc(doc(db, `liveSessions/${sessionId}/participants`, userId), {
    userId,
    name,
    avatarUrl: avatarUrl || null,
    role: 'student',
    isOnline: true,
    joinedAt: now,
    lastActiveAt: now,
  });
  await updateParticipantCount(sessionId);
};

export const leaveLiveSession = async (sessionId: string, userId: string): Promise<void> => {
  await deleteDoc(doc(db, `liveSessions/${sessionId}/participants`, userId));
  await updateParticipantCount(sessionId);
};

export const kickParticipant = async (sessionId: string, userId: string): Promise<void> => {
  await leaveLiveSession(sessionId, userId);
};

// ---- Real-time Subscriptions ----

export const subscribeToLiveSession = (
  sessionId: string,
  callback: (session: LiveSession | null, participants: LiveParticipant[]) => void
) => {
  let session: LiveSession | null = null;
  let participants: LiveParticipant[] = [];

  const unsubSession = onSnapshot(doc(db, 'liveSessions', sessionId), (docSnap) => {
    if (docSnap.exists()) {
      session = { id: docSnap.id, ...docSnap.data() } as LiveSession;
      callback(session, participants);
    } else {
      callback(null, []);
    }
  });

  const unsubParticipants = onSnapshot(
    collection(db, `liveSessions/${sessionId}/participants`),
    (snap) => {
      participants = snap.docs.map(d => ({ ...d.data() } as LiveParticipant));
      if (session) callback(session, participants);
    }
  );

  return () => {
    unsubSession();
    unsubParticipants();
  };
};

export const subscribeToAnnotations = (
  sessionId: string,
  callback: (annotations: LiveAnnotation[]) => void
) => {
  const q = query(
    collection(db, `liveSessions/${sessionId}/annotations`),
    orderBy('createdAt', 'asc')
  );

  return onSnapshot(q, (snap) => {
    const annotations = snap.docs.map(d => ({ id: d.id, ...d.data() } as LiveAnnotation));
    callback(annotations);
  });
};

export const subscribeToReactions = (
  sessionId: string,
  callback: (reactions: LiveReaction[]) => void
) => {
  const q = query(
    collection(db, `liveSessions/${sessionId}/reactions`),
    orderBy('createdAt', 'desc'),
    limit(50)
  );

  return onSnapshot(q, (snap) => {
    const reactions = snap.docs.map(d => ({ id: d.id, ...d.data() } as LiveReaction));
    callback(reactions);
  });
};

// ---- Annotations (Drawing / Laser) ----

export const addAnnotation = async (
  sessionId: string,
  annotation: Omit<LiveAnnotation, 'id' | 'createdAt'>
): Promise<string> => {
  const docRef = await addDoc(collection(db, `liveSessions/${sessionId}/annotations`), {
    ...annotation,
    createdAt: new Date().toISOString(),
  });
  return docRef.id;
};

export const clearAnnotations = async (sessionId: string): Promise<void> => {
  const snap = await getDocs(collection(db, `liveSessions/${sessionId}/annotations`));
  const batch = writeBatch(db);
  snap.docs.forEach(d => batch.delete(d.ref));
  await batch.commit();
};

// ---- Cursor Updates (Throttled) ----

let cursorTimeout: ReturnType<typeof setTimeout> | null = null;
let pendingCursor: { x: number; y: number } | null = null;

export const updateTeacherCursor = (
  sessionId: string,
  userId: string,
  x: number,
  y: number
): void => {
  pendingCursor = { x, y };

  if (cursorTimeout) return; // already scheduled

  cursorTimeout = setTimeout(async () => {
    if (pendingCursor) {
      try {
        await updateDoc(doc(db, `liveSessions/${sessionId}/participants`, userId), {
          cursorX: pendingCursor.x,
          cursorY: pendingCursor.y,
          lastActiveAt: new Date().toISOString(),
        });
      } catch (e) {
        console.warn('Failed to update cursor:', e);
      }
    }
    cursorTimeout = null;
    pendingCursor = null;
  }, 100); // 100ms throttle
};

export const clearCursorThrottle = (): void => {
  if (cursorTimeout) {
    clearTimeout(cursorTimeout);
    cursorTimeout = null;
  }
  pendingCursor = null;
};

// ---- Reactions ----

export const addReaction = async (
  sessionId: string,
  userId: string,
  userName: string,
  type: LiveReaction['type']
): Promise<void> => {
  await addDoc(collection(db, `liveSessions/${sessionId}/reactions`), {
    sessionId,
    userId,
    userName,
    type,
    createdAt: new Date().toISOString(),
  });
};
